woocommerce/woocommerce-ios
# Reviews: Fix missing product information on first load

Closes: #1907 


### Description

Previously we were synchronizing reviews by concurrently fetching reviews, products, and notifications and waiting until all these requests succeed to complete the synching.

Fetching products requires a list of product IDs from the fetched reviews. So there's a potential issue: if the request to fetch products starts before the request to fetch reviews completes, we'll send a request to fetch products with an empty ID list. If we haven't fetched products for the store before, we'll end up with reviews without product information.

The solution is to only sync products after the review synching is done. Required product IDs are taken from the fetched reviews and sent directly to the product synching request. This leads to the need to change the return result of `synchronizeProductReviews` to the modern `Result` with a list of reviews (hence changes in Yosemite and Networking layers).

### Testing instructions

- Log out from your account and log back in.
- Switch immediately to the Menu tab and select Reviews.
- Notice that after loading is done, all the reviews have correct product information.

Note: you may find that the first fetch will take quite a while. We'll have to look into why it takes so long and optimize it.

### Screenshots

https://user-images.githubusercontent.com/5533851/156700129-872ac2a1-bbc0-43f0-9c65-4f0548ed624c.mp4


---
